WireDomain/Tests/WireDomainTests/Repositories/ConversationLabelsRepositoryTests.swift

```diff
@@ -0,0 +1,332 @@
+//
+
+@testable import WireAPI
+import WireAPISupport
+@testable import WireDataModel
+import WireDataModelSupport
+@testable import WireDomain
+import XCTest
+
+final class ConversationLabelsRepositoryTests: XCTestCase {
+
+    var sut: ConversationLabelsRepository!
+    var userPropertiesAPI: MockUserPropertiesAPI!
+
+    var stack: CoreDataStack!
+    let coreDataStackHelper = CoreDataStackHelper()
+    let modelHelper = ModelHelper()
+
+    private var conversation1: ZMConversation!
+    private var conversation2: ZMConversation!
+    private var conversation3: ZMConversation!
+
+    var context: NSManagedObjectContext {
+        stack.syncContext
+    }
+
+    override func setUp() async throws {
+        try await super.setUp()
+        /// Batch requests don't work with in-memory store
+        /// so we need to use a persistent store.
+        stack = try await coreDataStackHelper.createStack(inMemoryStore: false)
+        cleanLabelEntity(in: context)
+        setupConversations()
+        userPropertiesAPI = MockUserPropertiesAPI()
+        sut = ConversationLabelsRepository(
+            userPropertiesAPI: userPropertiesAPI,
+            context: context
+        )
+    }
+
+    override func tearDown() async throws {
+        stack = nil
+        userPropertiesAPI = nil
+        sut = nil
+        try await super.tearDown()
+    }
+
+    // MARK: - Tests
+
+    func testPullConversation_Given_Local_Store_Empty_Labels_Are_Created() async throws {
+        // Mock
+
+        userPropertiesAPI.getPropertyForKey_MockValue = .conversationLabels([
+            Scaffolding.conversationLabel1,
+            Scaffolding.conversationLabel2
+        ])
+
+        // When
+
+        try await sut.pullConversationLabels()
+
+        // Then
+
+        try await context.perform { [context] in
+            let fetchRequest = NSFetchRequest<Label>(entityName: Label.entityName())
+            let a = try context.fetch(fetchRequest)
+            let remoteIdentifiers = a.map(\.remoteIdentifier)
+            XCTAssert(remoteIdentifiers.contains(Scaffolding.conversationLabel1.id))
+            XCTAssert(remoteIdentifiers.contains(Scaffolding.conversationLabel2.id))
+        }
+    }
+
+    func testPullConversationLabels_Given_Label_Exist_Locally_Label_Name_Is_Updated() async throws {
+        // Given
+
+        _ = await context.perform { [context] in
+            var created = false
+            let label = Label.fetchOrCreate(remoteIdentifier: Scaffolding.conversationLabel1.id, create: true, in: context, created: &created)
+            label?.name = Scaffolding.conversationLabel1.name
+            context.saveOrRollback()
+        }
+
+        try await context.perform { [context] in
+            let fetchRequest = NSFetchRequest<Label>(entityName: Label.entityName())
+            let a = try context.fetch(fetchRequest)
+            XCTAssertNotNil(a.first)
+            XCTAssertEqual(a.first!.name, Scaffolding.conversationLabel1.name)
```

Comment:
I think this will crash the test if there is no first. You could use `try AssertCount(a, count: 1)` which will throw if the count is wrong.

Reply:
I am using `XCTAssertNotNil(a.first)` to make sure of that.

Reply:
I think @johnxnguyen's point was that even if `XCTAssertNotNil` fails the test will continue so it will crash on `XCTAssertEqual`. `AssertCount` is our custom matcher which will exit early so to prevent following crashes

Reply:
you're right, I will change that, thx

Reply:
done